great_ape_safe/ape_api/euler.py

```diff
@@ -0,0 +1,49 @@
+from brownie import interface
+from helpers.addresses import r
+from helpers.constants import MaxUint256
+
+
+class Euler:
+    def __init__(self, safe):
+        self.safe = safe
+        self.euler = safe.contract(r.euler.euler)
+        self.markets = interface.IEuler(r.euler.euler_markets, owner=safe.account)
+
+    def deposit(self, underlying, amount):
+        etoken = interface.IEToken(
+            self.markets.underlyingToEToken(underlying), owner=self.safe.account
+        )
+
+        before_etoken = etoken.balanceOf(self.safe)
+        before_underlying = underlying.balanceOf(self.safe)
+
+        underlying.approve(self.euler, amount)
+        etoken.deposit(0, amount)
+
+        after_etoken = etoken.balanceOf(self.safe)
+        after_underlying = underlying.balanceOf(self.safe)
+
+        assert after_etoken > before_etoken
+        assert after_underlying < before_underlying
+
+    def deposit_all(self, underlying):
+        self.deposit(underlying, underlying.balanceOf(self.safe))
+
+    def withdraw(self, underlying, amount):
+        etoken = interface.IEToken(
+            self.markets.underlyingToEToken(underlying), owner=self.safe.account
+        )
+
+        before_etoken = etoken.balanceOf(self.safe)
+        before_underlying = underlying.balanceOf(self.safe)
+
+        etoken.withdraw(0, amount)
+
+        after_etoken = etoken.balanceOf(self.safe)
+        after_underlying = underlying.balanceOf(self.safe)
+
+        assert after_etoken < before_etoken
+        assert after_underlying > before_underlying
+
+    def withdraw_all(self, underlying):
+        self.withdraw(underlying, MaxUint256)
```

Comment:
nice, confirming the logic: [ref](https://github.com/euler-xyz/euler-contracts/blob/cd3036e0087280365819f99ad531141894d0b7ee/contracts/BaseLogic.sol#L394)